internal/fxlog/fxlog.go

```diff
@@ -21,86 +21,145 @@
 package fxlog
 
 import (
+	"os"
+	"strings"
+
+	"go.uber.org/fx/internal/fxreflect"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
-// Level is the level of logging used by Logger.
-type Level int
+// Logger defines interface used for logging.
+type Logger interface {
+	// LogEvent is called when a logging event is emitted.
+	LogEvent(Event)
+}
 
-const (
-	// InfoLevel is Info logging level used to log messages via zap.
-	InfoLevel Level = iota
-	// ErrorLevel is Info logging level used to log messages via zap.
-	ErrorLevel
-)
+// Event defines an event emitted by fx.
+type Event interface {
+	event() // Only fxlog can implement this interface.
+}
 
-// Entry is an entry to be later serialized into zap message and fields.
-type Entry struct {
-	Level   Level
-	Message string
-	Fields  []Field
-	Stack   string
+// Passing events by type to make Event hashable in the future.
+func (*LifecycleOnStartEvent) event() {}
+func (*LifecycleOnStopEvent) event()  {}
+func (*ApplyOptionsError) event()     {}
+func (*SupplyEvent) event()           {}
+func (*ProvideEvent) event()          {}
+func (*InvokeEvent) event()           {}
+func (*InvokeFailedEvent) event()     {}
+func (*StartFailureError) event()     {}
+func (*StopSignalEvent) event()       {}
+func (*StopErrorEvent) event()        {}
+func (*StartErrorEvent) event()       {}
+func (*StartRollbackError) event()    {}
+func (*RunningEvent) event()          {}
+
+// LifecycleOnStartEvent is emitted for whenever an OnStart hook is executed
+type LifecycleOnStartEvent struct {
```

Comment:
suggestion: LifecycleHookStartEvent (since the thing you put into the lifecycle is called a Hook)